tests/src/Kernel/Controller/WopiControllerTest.php

```diff
@@ -0,0 +1,432 @@
+    $request = $this->createRequest(write: TRUE);
+    $assert_response($request, TRUE);
+  }
+
+  /**
+   * Tests successful requests to the 'collabora-online.wopi.contents' route.
+   *
+   * @covers ::wopiGetFile
+   */
+  public function testWopiGetFile(): void {
+    $request = $this->createRequest(
+      '/contents',
+    );
+    $response = $this->handleRequest($request);
+    $this->assertEquals(Response::HTTP_OK, $response->getStatusCode());
+    $this->assertEquals($this->file->getFileUri(), $response->getFile());
+    $this->assertEquals($this->file->getMimeType(), $response->headers->get('Content-Type'));
+  }
+
+  /**
+   * Tests successful requests to the 'collabora-online.wopi.save' route.
+   *
+   * @covers ::wopiPutFile
+   */
+  public function testWopiPutFile(): void {
+    $file_changed_time = date_create_immutable_from_format('U', (string) $this->file->getChangedTime());
+    $assert_response = function (Request $request, string $log_message) use ($file_changed_time): void {
+      $this->logger->reset();
+      $this->assertResponse(
+        Response::HTTP_OK,
+        json_encode([
+          'LastModifiedTime' => $file_changed_time->format('c'),
+        ]),
+        'application/json',
+        $request
+      );
+      $log_message ??= 'Save reason: Saved by Collabora Online';
+      $this->assertTrue($this->logger->hasRecord($log_message), 'error');
+    };
+
+    // Test a successful save request without timestamp header.
+    $request = $this->createRequest('/contents', 'POST', write: TRUE);
+    $log_message = 'Save reason: Saved by Collabora Online';
+    $assert_response($request, $log_message);
+
+    // Test a successful save request with a timestamp header.
+    $request->headers->set('x-cool-wopi-timestamp', $file_changed_time->format(\DateTimeInterface::ATOM));
+    $assert_response($request, $log_message);
+
+    // Test how different headers result in different log messages.
+    $request->headers->set('x-cool-wopi-ismodifiedbyuser', 'true');
+    $log_message = 'Save reason: Saved by Collabora Online (Modified by user)';
+    $assert_response($request, $log_message);
+
+    $request->headers->set('x-cool-wopi-isautosave', 'true');
+    $log_message = 'Save reason: Saved by Collabora Online (Modified by user, Autosaved)';
+    $assert_response($request, $log_message);
+
+    $request->headers->set('x-cool-wopi-isexitsave', 'true');
+    $log_message = 'Save reason: Saved by Collabora Online (Modified by user, Autosaved, Save on Exit)';
+    $assert_response($request, $log_message);
+  }
+
+  /**
+   * Tests the 'collabora-online.wopi.save' route with a conflicting timestamp.
+   *
+   * @covers ::wopiPutFile
+   */
+  public function testWopiPutFileConflict(): void {
+    $request = $this->createRequest('/contents', 'POST', write: TRUE);
+
+    // Set a time in the future to force the error.
+    $file_changed_time = date_create_immutable_from_format('U', (string) ($this->file->getChangedTime() + 1000));
```

Comment:
Should we use the method from the immutable class instead of the function?

Reply:
I took this from the previous version of these tests in this PR